### PR TITLE
 Implement dark mode on learn area landing page

### PR DESCRIPTION
**Pull request by oyenuga17**
@SaySayo @sabine @punchagan please review.
- To test this uncomment line https://github.com/oyenuga17/ocaml.org/blob/14e93c389c0921d0ff9eaa1ae5ba2149a7862206/src/ocamlorg_frontend/layouts/layout.eml#L103 in layout.eml

### CODE REVIEW

File: src/ocamlorg_frontend/pages/learn.eml [hunk 1]
```diff
@@ -8,12 +8,12 @@ Learn_layout.single_column_layout
 ~canonical:Url.learn
 ~active_top_nav_item:Header.Learn
 ~current:Overview @@
-  <div class="w-full border-b bg-[#DDEAFC] md:bg-[length:100%_100%] md:bg-transparent md:bg-[url('/img/learn/hero-background.svg')]">
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml:11]: background color cannot be removed for the small screens, see the mobile design, there's a solid color background both in light mode and in dark mode

File: src/ocamlorg_frontend/pages/learn.eml [hunk 2]
```diff
@@ -22,35 +22,38 @@ Learn_layout.single_column_layout
...
-        <span class="hidden lg:inline">Jump to</span> Intermediate Section
-        <%s! Learn_components.little_down_arrow_icon %>
-      </a>
-      <a href="#advanced_section" class="py-3 lg:py-6 flex items-center gap-2 text-primary font-bold hover:underline">
-        <%s! Learn_components.advanced_section_icon %>
-        <span class="hidden lg:inline">Jump to</span> Advanced Section
-        <%s! Learn_components.little_down_arrow_icon %>
-      </a>
+  <div class="w-full bg:white dark:bg-dark-background_jumping_section">
+    <div class="container-fluid bg:white dark:bg-dark-background_jumping_section">
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]: Setting background color twice is unnecessary here. Color can be set on the outer container.

File: src/ocamlorg_frontend/pages/learn.eml [hunk 3]
```diff
@@ -22,35 +22,38 @@ Learn_layout.single_column_layout
...
+        </div>
     </div>
-
+  </div>
+  <div class="container-fluid">
     <h2 id="beginner_section" class="mt-16 mb-8 text-text_content text-3xl font-regular leading-9 tracking-widest">
-      <span class="text-tertiary font-bold">FOUNDATIONS</span>
-      FOR BEGINNERS
+      <span class="text-secondary font-bold">FOUNDATIONS</span>
+      <span class="dark:text-dark-text">FOR BEGINNERS</span>
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]: shouldn't introduce a span here, but instead set the text color on the surrounding `h2` just like it is done for light mode

File: src/ocamlorg_frontend/pages/learn.eml [hunk 4]
```diff
@@ -101,7 +104,7 @@ Learn_layout.single_column_layout
     </div>
   </div>
 
-  <div class="section-mild-contrast">
+  <div class="section-mild-contrast dark:bg-none dark:bg-gradient-to-r dark:from-[#010105] dark:to-[#10162D] dark:border-none">
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]: Styles should be added to the existing `section-mild-contrast` class (see `bg_sections.css`), if Tailwind allows that. If Tailwind doesn't allow that, introduce a `dark:section-mild-contrast` class.

File: src/ocamlorg_frontend/pages/learn.eml [hunk 5]
```diff
@@ -145,7 +148,7 @@ Learn_layout.single_column_layout
           %>
     </div>
 
-    <div class="border-t my-6"></div>
+    <div class="border-t border-separator_30 dark:border-dark-separator_30  my-6"></div>
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]:
```suggestion
    <div class="border-t border-separator_30 dark:border-dark-separator_30 my-6"></div>
```

double space - unnecessary whitespace

File: src/ocamlorg_frontend/pages/learn.eml [hunk 6]
```diff
@@ -172,8 +175,8 @@ Learn_layout.single_column_layout
     <%s! Learn_components.lang_manual_banner %>
 
     <h2 id="advanced_section" class="mt-16 mb-8 text-text_content text-3xl font-regular leading-9 tracking-widest">
-      <span class="text-primary font-bold">ELEVATED</span>
-      EXPERTISE
+      <span class="text-primary dark:text-dark-primary font-bold">ELEVATED</span>
+      <span class="text_title dark:text-dark-text">EXPERTISE</span>
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]: same as earlier, text styles on `h2`

File: src/ocamlorg_frontend/pages/learn.eml [hunk 7]
```diff
@@ -193,15 +196,15 @@ Learn_layout.single_column_layout
         <% papers |> List.iter (fun (paper : Data.Paper.t) -> %>
         <a
           href="<%s paper.links |> List.hd |> (fun x -> x.uri) %>"
-          class="card p-6 flex-1"
+          class="card p-6 flex-1 dark:bg-dark-background_light dark:hover:bg-dark-window_code"
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]: card background hover color needs to be made a color variable in Figma, then we can use it here. This is definitely not the code window color (or that color is just badly named)

File: src/ocamlorg_frontend/pages/learn.eml [hunk 8]
```diff
@@ -193,15 +196,15 @@ Learn_layout.single_column_layout
...
         >
-          <div class="font-semibold text-base mb-3"><%s paper.title %></div>
-          <div style="display: -webkit-box;-webkit-line-clamp: 4;-webkit-box-orient: vertical;overflow: hidden;"><%s paper.abstract %></div>
-          <div class="font-bold text-sm mt-3"><%s String.concat ", " paper.authors %></div>
+          <div class="font-semibold text-base dark:text-dark-text mb-3"><%s paper.title %></div>
+          <div class="dark:text-dark-grey_navigation" style="display: -webkit-box;-webkit-line-clamp: 4;-webkit-box-orient: vertical;overflow: hidden;"><%s paper.abstract %></div>
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]: `grey_navigation` color is going to be renamed / replaced in Figma

**Review by sabine — COMMENTED** [review]
Some great work, thanks @oyenuga17!

This uncovers that we need to make the Figma more consistent and improve the color names. :+1:

File: src/ocamlorg_frontend/pages/learn.eml (hunk 4 above)
Comment (oyenuga17) [reply src/ocamlorg_frontend/pages/learn.eml]: Hi @sabine  i am not sure how to apply these style in `bg_sections.css` i tried this https://github.com/oyenuga17/ocaml.org/blob/db689764c9cf2802241f4710bca2043a9fa0973a/src/ocamlorg_frontend/css/partials/bg_sections.css#L8-L11 but it didn't work.

Comment (sabine) [reply src/ocamlorg_frontend/pages/learn.eml]: When you do it like this in `bg_sections.css`, it works:

```
.dark .section-mild-contrast {
  background: linear-gradient(45deg, #010105, #10162D);
  @apply border-none;
}
```

File: src/ocamlorg_frontend/css/partials/bg_sections.css [hunk 9]
```diff
@@ -2,6 +2,11 @@
 
 .section-mild-contrast {
     background: url("/img/innerbottombg.png") left bottom no-repeat, url("/img/innertopbg.png") right top no-repeat;
-    @apply bg-legacy-mild-contrast border-b border-t;
+    @apply bg-sand border-b border-t;
+  }
+
+.dark-section-mild-contrast {
```
Comment (sabine) [inline src/ocamlorg_frontend/css/partials/bg_sections.css:8]:
```suggestion
.dark .section-mild-contrast {
```

File: src/ocamlorg_frontend/pages/learn.eml [hunk 10]
```diff
@@ -22,34 +22,37 @@ Learn_layout.single_column_layout
...
-      </a>
+  <div class="w-full bg:white dark:bg-dark-background_navigation">
+    <div class="container-fluid">
+      <div class="hidden md:grid md:gap-6 md:grid-cols-3">
+        <a href="#beginner_section" class="py-3 lg:py-6 flex items-center gap-2 text-secondary font-bold hover:underline">
+          <%s! Learn_components.beginner_section_icon %>
+          <span class="hidden lg:inline">Jump to</span> Beginner Section
+          <%s! Learn_components.little_down_arrow_icon %>
+        </a>
+        <a href="#intermediate_section" class="py-3 lg:py-6 flex items-center gap-2 text-tertiary font-bold hover:underline">
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]:
```suggestion
        <a href="#intermediate_section" class="py-3 lg:py-6 flex items-center gap-2 text-tertiary dark:text-dark-tertiary font-bold hover:underline">
```

File: src/ocamlorg_frontend/pages/learn.eml [hunk 11]
```diff
@@ -22,34 +22,37 @@ Learn_layout.single_column_layout
...
-      </a>
-      <a href="#advanced_section" class="py-3 lg:py-6 flex items-center gap-2 text-primary font-bold hover:underline">
-        <%s! Learn_components.advanced_section_icon %>
-        <span class="hidden lg:inline">Jump to</span> Advanced Section
-        <%s! Learn_components.little_down_arrow_icon %>
-      </a>
+  <div class="w-full bg:white dark:bg-dark-background_navigation">
+    <div class="container-fluid">
+      <div class="hidden md:grid md:gap-6 md:grid-cols-3">
+        <a href="#beginner_section" class="py-3 lg:py-6 flex items-center gap-2 text-secondary font-bold hover:underline">
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]:
```suggestion
        <a href="#beginner_section" class="py-3 lg:py-6 flex items-center gap-2 text-secondary dark:text-dark-secondary font-bold hover:underline">
```

File: src/ocamlorg_frontend/pages/learn.eml [hunk 12]
```diff
@@ -22,34 +22,37 @@ Learn_layout.single_column_layout
...
+          <%s! Learn_components.little_down_arrow_icon %>
+        </a>
+        </div>
     </div>
-
-    <h2 id="beginner_section" class="mt-16 mb-8 text-text_content text-3xl font-regular leading-9 tracking-widest">
-      <span class="text-tertiary font-bold">FOUNDATIONS</span>
+  </div>
+  <div class="container-fluid">
+    <h2 id="beginner_section" class="mt-16 mb-8 text-content text-3xl font-regular leading-9 tracking-widest">
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]:
```suggestion
    <h2 id="beginner_section" class="mt-16 mb-8 text-content dark:text-dark-title text-3xl font-regular leading-9 tracking-widest">
```

File: src/ocamlorg_frontend/pages/learn.eml [hunk 13]
```diff
@@ -112,13 +115,13 @@ Learn_layout.single_column_layout
...
-      <a class="pt-6 pb-3 text-primary hover:underline" href="<%s Url.platform %>">Go to Platform Tools Documentation</a>
+      <a class="pt-6 pb-3 text-primary dark:text-dark-primary dark:underline hover:underline" href="<%s Url.platform %>">Go to Platform Tools Documentation</a>
     </div>
   </div>
 
   <div class="container-fluid">
-    <h2 id="intermediate_section" class="mt-16 mb-8 text-text_content text-3xl font-regular leading-9 tracking-widest">
-      <span class="text-[#3882B7] font-bold">EXPLORING</span>
+    <h2 id="intermediate_section" class="mt-16 mb-8 text-content dark:text-dark-title text-3xl font-regular leading-9 tracking-widest">
+      <span class="text-tertiary font-bold">EXPLORING</span>
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]:
```suggestion
      <span class="text-tertiary dark:text-dark-tertiary font-bold">EXPLORING</span>
```

File: src/ocamlorg_frontend/pages/learn.eml [hunk 14]
```diff
@@ -171,9 +174,9 @@ Learn_layout.single_column_layout
 
     <%s! Learn_components.lang_manual_banner %>
 
-    <h2 id="advanced_section" class="mt-16 mb-8 text-text_content text-3xl font-regular leading-9 tracking-widest">
-      <span class="text-primary font-bold">ELEVATED</span>
-      EXPERTISE
+    <h2 id="advanced_section" class="mt-16 mb-8 text-content text-3xl font-regular leading-9 tracking-widest">
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]:
```suggestion
    <h2 id="advanced_section" class="mt-16 mb-8 text-content dark:text-dark-content text-3xl font-regular leading-9 tracking-widest">
```

File: src/ocamlorg_frontend/pages/learn.eml [hunk 15]
```diff
@@ -183,8 +186,8 @@ Learn_layout.single_column_layout
     </div>
 
     <div class="mt-10 lg:mt-20">
-      <h3 class="font-bold">Papers</h3>
-      <div class="mt-6 text-legacy-lighter text-lg">
+      <h3 class="font-bold dark:text-dark-title">Papers</h3>
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]:
```suggestion
      <h3 class="font-bold text-title dark:text-dark-title">Papers</h3>
```

File: src/ocamlorg_frontend/pages/learn.eml [hunk 16]
```diff
@@ -193,15 +196,15 @@ Learn_layout.single_column_layout
         <% papers |> List.iter (fun (paper : Data.Paper.t) -> %>
         <a
           href="<%s paper.links |> List.hd |> (fun x -> x.uri) %>"
-          class="card p-6 flex-1"
+          class="card p-6 flex-1 dark:bg-dark-card dark:hover:bg-dark-card_hover"
```
Comment (sabine) [inline src/ocamlorg_frontend/pages/learn.eml]: same as with the background sections, apply styles in `cards.css`.

```suggestion
          class="card p-6 flex-1"
```

File: src/ocamlorg_frontend/css/partials/cards.css [hunk 17]
```diff
@@ -7,5 +7,5 @@
 }
 
 .card {
-  @apply bg-white drop-shadow hover:shadow-[0px_0px_12px_0px_#d5d5d5] rounded-2xl border;
+  @apply bg-white drop-shadow hover:shadow-[0px_0px_12px_0px_#00000080] rounded-2xl border;
```
Comment (sabine) [inline src/ocamlorg_frontend/css/partials/cards.css]:
```suggestion
  @apply bg-white drop-shadow hover:shadow-[0px_0px_12px_0px_#00000080] rounded-2xl border;
}

.dark-card {
  @apply bg-dark-card drop-shadow hover:shadow-none rounded-2xl border;
```

We can use `card dark:dark-card` classes when we wrap these utilities with `@layer utilities` like described here: https://stackoverflow.com/questions/75889347/how-can-i-use-a-custom-class-with-tailwind-css-for-dark-mode

File: src/ocamlorg_frontend/layouts/learn_layout.eml [hunk 18]
```diff
@@ -51,13 +51,13 @@ let tabs
...
             </select>
-            <span class="text-orange-700 cursor-pointer">&#x25BE;</span>
+            <span class="text-primary dark:text-dark-primary cursor-pointer">&#x25BE;</span>
         </li>
       </ul>
     </nav>
   in
     <%s! breadcrumb_options current %>
-    <div class="bg-legacy-text-title hidden md:flex">
+    <div class="bg-title hidden md:flex">
```
Comment (sabine) [inline src/ocamlorg_frontend/layouts/learn_layout.eml]: no dark background color is being set here

File: src/ocamlorg_frontend/css/partials/bg_sections.css [hunk 19]
```diff
@@ -2,6 +2,11 @@
...
+  background: linear-gradient(45deg, #010105, #10162D);
+    @apply border-none;
```
Comment (sabine) [inline src/ocamlorg_frontend/css/partials/bg_sections.css]: We need to wrap the file content in `@layer utilities`, then the `dark:dark-section-milt-contrast` class will work as expected.

File: tailwind.config.js [hunk 20]
```diff
@@ -7,41 +7,48 @@ const figma_colors = {
...
+      tertiary: "#007AD0",
       tertiary_lighter: "#007BC7",
       tertiary_bt_hover: "#111827",
-      tertiary_bt_pressed: "#020C22",
 
-      text: "#FFFFFF",
-      grey_navigation: "#868686",
+      title: "#FFFFFFDE",
+      content: "FFFFFF99",
+      card_hover: "2B2A2A",
```
Comment (sabine) [inline tailwind.config.js]:
```suggestion
      content: "#FFFFFF99",
      card_hover: "#2B2A2A",
```

**Review by sabine — COMMENTED** [review]
I just learned something about Tailwind and classes that live in css files, see below for explanation and stack overflow link!


Almost good to merge!

**Comment by oyenuga17** [comment]
@sabine please help check again I have made all the requested changes. I also learnt @layers concept in tailwind 😁, thanks for pointing out that.

**Comment by oyenuga17** [comment]
@Clairevanden  I have attended to all the comments on figma. @sabine can you please review and make this available in staging so Claire can also review. 🚀

**Comment by Clairevanden** [comment]
Awesome :) Thanks Seun

**Comment by sabine** [comment]
Ideally @SaySayo and @punchagan will have a look at the changes and the Figma document to see if they can find mistakes. (It's also going to be helpful for them review the previous review comments to get a better understanding of how dark mode and specifically layer classes work in Tailwind.)

I can put this on staging again later today.

I think we'll have to put these pages on staging in a way such that the styles do not affect other pages (specifically the documentation pages). So, for the sake of being able to put this on staging, it probably makes sense to explicitly activate the dark mode only for those pages which have a fully working dark mode. (Which could mean moving the dark mode script to only the page that you're currently implementing.)

**Comment by oyenuga17** [comment]
Hi @sabine this make sense, i think another way to go about this is manually inserting this script `document.documentElement.classList.add('dark')` and `document.documentElement.classList.remove('dark')`  into the browser console any time we want to test dark mode instead of putting the script in individual pages and having to remove later when the dark mode is fully ready.

**Comment by sabine** [comment]
Yes, that's another way to do it that would not rely on removing the script
before merging the PR.

**Pull request merged**